jf_agent/agent_logging.py

```diff
@@ -1,50 +1,91 @@
-from datetime import datetime
 from contextlib import contextmanager
 from functools import wraps
 import logging
 import os
-import traceback
+import sys
+import urllib3
 
 '''
 Guidance on logging/printing in the agent:
 
-stdout (e.g., print() output) is for the customer/operator; it should include
- - info on run progress: tqdm output, etc.
- - errors/warnings encountered
+stdout (e.g., print() output) is for the customer/operator; it should be used when
+we have customer sensitive data that they do not want to submit to jellyfish (i.e. passwords)
 
-logger output is for Jellyfish; it should include
+logger output is for Jellyfish and customers; it should include:
  - function entry/exit timings
  - loop iteration timings/liveness indicators
  - errors/warnings encountered
 
 logger output is submitted to Jellyfish, so should only contain non-sensitive data:
 e.g., function names, iteration counts
-
-Since we generally want errors/warnings to go to BOTH stdout and the logger, we
-should generally use log_and_print() instead of logger.whatever().
 '''
 
 LOG_FILE_NAME = 'jf_agent.log'
 
+# For styling in log files, I think it's best to always use new lines even when we use
+# the special character to ignore them. Leverage always_use_newlines for this
+def _emit_helper(_self: logging.Handler, record: logging.LogRecord, always_use_newlines=False):
+    special_code = '[!n]'
```

Comment:
Q: When would we want to ignore newlines in a message? Why was this functionality added?

Reply:
This is recreating legacy functionality for a good UX experience. For Jira download we used to have out put like this:
```
Obtained Jira configuration, attempting download...
downloading jira fields... ✓
downloading jira projects... ✓
downloading jira project components... ✓
downloading jira versions... ✓
downloading jira users...  ✓
downloading jira resolutions... ✓
downloading jira issue types...  ✓
downloading jira issue link types... ✓
downloading jira priorities... ✓
```

I opted to leave it in for Jira because most of Jira doesn't use TQDM progress bars, so the check marks make sense. I got rid of it for most git tasks (like downloading PRs) because they all use TQDM which mudies up the readability of the check marks imo

Reply:
We DO NOT want to print a new line here because we need to print `downloading jira fields... `, run some code, and then print the check mark inline with the previous print statement